src/scoped.rs

```diff
@@ -146,47 +144,61 @@ pub unsafe fn builder_spawn_unsafe<'a, F>(
 where
     F: FnOnce() + Send + 'a,
 {
-    use std::mem;
-
-    let closure: Box<FnBox + 'a> = Box::new(f);
-    let closure: Box<FnBox + Send> = mem::transmute(closure);
+    let closure: Box<FnBox<()> + 'a> = Box::new(f);
+    let closure: Box<FnBox<()> + Send> = mem::transmute(closure);
     builder.spawn(move || closure.call_box())
 }
 
-
 pub struct Scope<'a> {
-    dtors: RefCell<Option<DtorChain<'a>>>,
+    /// The list of the deferred functions and thread join jobs.
+    dtors: RefCell<Option<DtorChain<'a, ()>>>,
+    // !Send + !Sync
+    _marker: PhantomData<*const ()>,
 }
 
-struct DtorChain<'a> {
-    dtor: Box<FnBox + 'a>,
-    next: Option<Box<DtorChain<'a>>>,
+struct DtorChain<'a, T> {
+    dtor: Box<FnBox<T> + 'a>,
+    next: Option<Box<DtorChain<'a, T>>>,
 }
 
-enum JoinState {
-    Running(thread::JoinHandle<()>),
-    Joined,
+impl<'a, T> DtorChain<'a, T> {
+    pub fn pop(chain: &mut Option<DtorChain<'a, T>>) -> Option<Box<FnBox<T> + 'a>> {
+        chain.take().map(|mut node| {
+            *chain = node.next.take().map(|b| *b);
+            node.dtor
+        })
+    }
 }
 
-impl JoinState {
-    fn join(&mut self) {
-        let mut state = JoinState::Joined;
-        mem::swap(self, &mut state);
-        if let JoinState::Running(handle) = state {
-            let res = handle.join();
+struct JoinState<T> {
+    join_handle: thread::JoinHandle<()>,
+    result: usize,
+    _marker: PhantomData<T>,
+}
 
-            if !thread::panicking() {
-                res.unwrap();
-            }
+impl<T: Send> JoinState<T> {
+    fn new(join_handle: thread::JoinHandle<()>, result: usize) -> JoinState<T> {
+        JoinState {
+            join_handle: join_handle,
+            result: result,
+            _marker: PhantomData,
         }
     }
+
+    fn join(self) -> thread::Result<T> {
+        let result = self.result;
+        self.join_handle.join().map(|_| {
+            unsafe { *Box::from_raw(result as *mut T) }
+        })
+    }
 }
 
 /// A handle to a scoped thread
-pub struct ScopedJoinHandle<T> {
-    inner: Rc<RefCell<JoinState>>,
-    packet: Arc<AtomicOption<T>>,
+pub struct ScopedJoinHandle<'a, T: 'a> {
```

Comment:
Do we still need the `'a` lifetime? I don't see how it's useful at the moment. :)